binaries/daemon/src/pending.rs

```diff
@@ -0,0 +1,164 @@
+            waiting_subscribers: HashMap::new(),
+            exited_before_subscribe: HashSet::new(),
+            reported_init_to_coordinator: false,
+        }
+    }
+
+    pub fn insert(&mut self, node_id: NodeId) {
+        self.local_nodes.insert(node_id);
+    }
+
+    pub fn set_external_nodes(&mut self, value: bool) {
+        self.external_nodes = value;
+    }
+
+    pub async fn handle_node_subscription(
+        &mut self,
+        node_id: NodeId,
+        reply_sender: oneshot::Sender<DaemonReply>,
+        coordinator_connection: &mut Option<TcpStream>,
+    ) -> eyre::Result<DataflowStatus> {
+        self.waiting_subscribers
+            .insert(node_id.clone(), reply_sender);
+        self.local_nodes.remove(&node_id);
+
+        self.update_dataflow_status(coordinator_connection).await
+    }
+
+    pub async fn handle_node_stop(
+        &mut self,
+        node_id: &NodeId,
+        coordinator_connection: &mut Option<TcpStream>,
+    ) -> eyre::Result<()> {
+        if self.local_nodes.remove(node_id) {
+            tracing::warn!("node `{node_id}` exited before initializing dora connection");
+            self.exited_before_subscribe.insert(node_id.clone());
+            self.update_dataflow_status(coordinator_connection).await?;
+        }
+        Ok(())
+    }
+
+    pub async fn handle_external_all_nodes_ready(&mut self, success: bool) -> eyre::Result<()> {
+        if !self.local_nodes.is_empty() {
+            bail!("received external `all_nodes_ready` event before local nodes were ready");
+        }
+        let external_error = if success {
+            None
+        } else {
+            Some("some nodes failed to initalize on remote machines".to_string())
+        };
+        self.answer_subscribe_requests(external_error).await;
+
+        Ok(())
+    }
+
+    async fn update_dataflow_status(
+        &mut self,
+        coordinator_connection: &mut Option<TcpStream>,
+    ) -> eyre::Result<DataflowStatus> {
+        if self.local_nodes.is_empty() {
+            if self.external_nodes {
+                if !self.reported_init_to_coordinator {
+                    self.report_nodes_ready(coordinator_connection).await?;
+                    self.reported_init_to_coordinator = true;
+                }
+                Ok(DataflowStatus::Pending)
+            } else {
+                self.answer_subscribe_requests(None).await;
+                Ok(DataflowStatus::AllNodesReady)
+            }
+        } else {
+            Ok(DataflowStatus::Pending)
+        }
+    }
+
+    async fn answer_subscribe_requests(&mut self, external_error: Option<String>) {
+        let result = if self.exited_before_subscribe.is_empty() {
+            match external_error {
+                Some(err) => Err(err),
+                None => Ok(()),
+            }
+        } else {
+            Err(format!(
+                "Nodes failed before subscribing: {:?}",
+                self.exited_before_subscribe
+            ))
```

Comment:
So, I tried with the python operator example, and if the webcam operator initialization fails, the other operators will also fail with error:

- `operator plot`:
```
   1 │ Traceback (most recent call last):
   2 │   File "<string>", line 1, in <module>
   3 │ RuntimeError: Dora Runtime raised an error.
   4 │ 
   5 │ Caused by:
   6 │    0: main task failed
   7 │    1: failed to init event stream
   8 │    2: subscribe failed: Nodes failed before subscribing: {NodeId("webcam")}
   9 │ 
  10 │ Location:
  11 │     apis/rust/node/src/event_stream/mod.rs:70:17
```

I think that a maybe more insightful error such as:
- ``webcam` node exited before subscribing to dora. However it is required for this node to run` 
- ``webcam` node exited before subscribing to dora. Please check: dora logs {dataflow_id} {<Failed node_id>}

The idea is to direct the debugging on the failing node